En primer lugar, vamos a realizar las importaciones necesarias para poder empezar a trabajar: 

In [2]:
#Importamos pandas
import pandas as pd 
#Importamos numpy
import numpy as np 
#Importamos nuestro archivo de soporte
import sys
sys.path.append('../')
from Jupyters.soporte_def import *

print ("Librerias importadas")

Librerias importadas


Una vez importadas las librerías de python necesarias, procedemos a importar el dataset original con el que trabajaremos y a hacer una copia del mismo para no modificar los datos reales.

In [3]:
#En primer lugar importamos el dataset
df_original = pd.read_csv ("../Clean_Data/Dataframe_original.csv")
print ("Dataframe importado")

Dataframe importado


In [4]:
#A continuación, vamos a copiar el dataframe a uno sobre el cuál vamos a trabajar
df_work = df_original
print ("Dataframe copiado")

Dataframe copiado


Una vez realizadas las tareas iniciales, procedemos con la limpieza del dataframe. Durante la limpieza de nuestro dataframe final vamos a enfocarnos en primer lugar en entender el set de datos, las columnas que lo conforman y el tipo de datos que forman nuestro dataset. Para ello seguimos los siguientes pasos:

***Paso 1: Exploración + Limpieza inicial***

In [5]:
initial_exploration(df_work)

El dataframe tiene 110191 registros y 54 columnas

El Df está formado por las siguientes columnas:
Index(['game_id', 'id', 'jersey_num', 'team_id_x', 'team_city_x',
       'team_name_x', 'team_abbreviation_x', 'full_name', 'is_active',
       'player_name', 'season', 'round_number', 'round_pick', 'overall_pick',
       'draft_type', 'team_id_y', 'team_city_y', 'team_name_y',
       'team_abbreviation_y', 'organization', 'organization_type',
       'player_profile_flag', 'first_name', 'last_name', 'display_first_last',
       'display_last_comma_first', 'display_fi_last', 'player_slug',
       'birthdate', 'school', 'country', 'last_affiliation', 'height',
       'weight', 'season_exp', 'jersey', 'position', 'rosterstatus',
       'games_played_current_season_flag', 'team_id', 'team_name',
       'team_abbreviation', 'team_code', 'team_city', 'playercode',
       'from_year', 'to_year', 'dleague_flag', 'nba_flag', 'games_played_flag',
       'draft_year', 'draft_round', 'draft_number', 

En primer lugar, vamos a intentar entender cada una de las columnas que conforman este dataframe y ver para que sirven. Posteriormente, decidiremos si esa columna es de utilidad o no y de no serlo la eliminaremos. También cambiaremos los nombres de las columnas para poder identificarlas mejor durante el estudio.

*Players_general*

- game_id: Identificador único del partido
- id: Identificador único del jugador
- first_name_x: Nombre del jugador (ya eliminada)
- last_name_x: Apellido del jugador (ya eliminada)
- jersey_num: Número de camiseta del jugador en dicho partido
- team_id_x: Identificador único del equipo para el que jugaba el jugador
- team_city_x: Ciudad del equipo
- team_name_x: Nombre del equipo 
- team_abbreviation_x: Abreviatura del equipo 


Esta tabla parece informar de los mejores jugadores en cada partido en la NBA hasta 2020 por lo que dado que no estamos analizando información de partidos si no que de los jugadores podemos eliminar la columna game_id ya que no le daremos uso. También vamos a renombrar varias columnas para poder identificarlas a posteriori.

In [6]:
df_work = df_work.drop(columns= ['game_id'])
print ("Columna eliminada")

Columna eliminada


In [7]:
df_work = renombrar_columnas (df_work, 'team_id_x', 'general_team')
df_work = renombrar_columnas (df_work, 'team_city_x', 'general_city')
df_work = renombrar_columnas (df_work, 'team_name_x', 'general_team_name')
df_work = renombrar_columnas (df_work, 'team_abbreviation_x', 'general_team_abbv')
df_work.columns

Nombre de la columna cambiado de team_id_x a general_team
Nombre de la columna cambiado de team_city_x a general_city
Nombre de la columna cambiado de team_name_x a general_team_name
Nombre de la columna cambiado de team_abbreviation_x a general_team_abbv


Index(['id', 'jersey_num', 'general_team', 'general_city', 'general_team_name',
       'general_team_abbv', 'full_name', 'is_active', 'player_name', 'season',
       'round_number', 'round_pick', 'overall_pick', 'draft_type', 'team_id_y',
       'team_city_y', 'team_name_y', 'team_abbreviation_y', 'organization',
       'organization_type', 'player_profile_flag', 'first_name', 'last_name',
       'display_first_last', 'display_last_comma_first', 'display_fi_last',
       'player_slug', 'birthdate', 'school', 'country', 'last_affiliation',
       'height', 'weight', 'season_exp', 'jersey', 'position', 'rosterstatus',
       'games_played_current_season_flag', 'team_id', 'team_name',
       'team_abbreviation', 'team_code', 'team_city', 'playercode',
       'from_year', 'to_year', 'dleague_flag', 'nba_flag', 'games_played_flag',
       'draft_year', 'draft_round', 'draft_number', 'greatest_75_flag'],
      dtype='str')

*Players*

- id: Identificador único del jugador
- full_name: Nombre completo del jugador
- first_name_y: Nombre del jugador (ya eliminada)
- last_name_y: Apellido del jugador (ya eliminada)
- is_active: Indicador de si el jugador está en activo actualmente


Esta tabla informa de todos los jugadores históricos de la NBA su nombre, apellido, nombre completo y si el jugador está en activo o no. De esta tabla vamos a mantener todas las columnas y no requiere de ninguna transformación adicional en este momento.

*Draft History*

- id: Identificador único del jugador
- player_name: Nombre completo del jugador
- season: Temporada en la que fue seleccionado
- round_number: Número de ronda en la que fue seleccionado el jugador
- round_pick: Número de selección dentro de la ronda
- overall_pick: Número de selección global en el draft
- draft_type: Tipo de draft en el que fue seleccionado
- team_id_y: Identificador único del equipo que seleccionó al jugador
- team_city_y: Ciudad del equipo que seleccionó al jugador
- team_name_y: Nombre del equipo que seleccionó al jugador
- team_abbreviation_y: Abreviatura del equipo que seleccionó al jugador
- organization: Nombre de la organización de procedencia del jugador
- organization_type: Tipo de organización de procedencia 
- player_profile_flag: Indicador de si el jugador tiene perfil en la NBA


Esta tabla informa de datos de la historia del draft de la nba de esta tabla vamos a mantener todas las columnas salvo la última ya que desconocemos realmente la utilidad de dicha columna y no impacta en nuestro estudio. También, vamos a modificar ciertas columnas:

In [8]:
df_work = df_work.drop(columns= ['player_profile_flag'])
print ("Columna eliminada")

Columna eliminada


In [9]:
df_work = renombrar_columnas(df_work, 'team_id_y', 'draft_team')
df_work = renombrar_columnas(df_work, 'team_city_y', 'draft_city')
df_work = renombrar_columnas(df_work, 'team_name_y', 'draft_team_name')
df_work = renombrar_columnas(df_work, 'team_abbreviation_y', 'draft_team_abbrv')
df_work.columns

Nombre de la columna cambiado de team_id_y a draft_team
Nombre de la columna cambiado de team_city_y a draft_city
Nombre de la columna cambiado de team_name_y a draft_team_name
Nombre de la columna cambiado de team_abbreviation_y a draft_team_abbrv


Index(['id', 'jersey_num', 'general_team', 'general_city', 'general_team_name',
       'general_team_abbv', 'full_name', 'is_active', 'player_name', 'season',
       'round_number', 'round_pick', 'overall_pick', 'draft_type',
       'draft_team', 'draft_city', 'draft_team_name', 'draft_team_abbrv',
       'organization', 'organization_type', 'first_name', 'last_name',
       'display_first_last', 'display_last_comma_first', 'display_fi_last',
       'player_slug', 'birthdate', 'school', 'country', 'last_affiliation',
       'height', 'weight', 'season_exp', 'jersey', 'position', 'rosterstatus',
       'games_played_current_season_flag', 'team_id', 'team_name',
       'team_abbreviation', 'team_code', 'team_city', 'playercode',
       'from_year', 'to_year', 'dleague_flag', 'nba_flag', 'games_played_flag',
       'draft_year', 'draft_round', 'draft_number', 'greatest_75_flag'],
      dtype='str')

*Common player info*

- id: Identificador único del jugador
- first_name: Nombre del jugador
- last_name: Apellido del jugador
- display_first_last: Nombre completo en formato Nombre Apellido
- display_last_comma_first: Nombre completo en formato Apellido, Nombre
- display_fi_last: Nombre completo en formato Inicial.Apellido
- player_slug: Identificador del jugador en formato URL
- birthdate: Fecha de nacimiento del jugador
- school:  Universidad o club de procedencia del jugador
- country: País de origen del jugador
- last_affiliation: Último equipo o institución a la que perteneció antes de la NBA + Selección Nacional 
- height: Altura del jugador
- weight: Peso del jugador
- season_exp: Años de experiencia en la NBA
- jersey: Número de camiseta del jugador
- position: Posición en la que juega el jugador
- rosterstatus: Jugador incluido en alguna plantilla actual (a fecha 2022)
- games_played_current_season_flag: Indicador de si el jugador ha disputado partidos en la temporada actual
- team_id: Identificador del equipo en el que más jugó (si inactivo) o equipo en el que juega actualmente.
- team_name: Nombre del equipo en el que más jugó (si inactivo) o equipo en el que juega actualmente.
- team_abbreviation: Abreviatura del equipo en el que más jugó (si inactivo) o equipo en el que juega actualmente.
- team_code: Código del equipo (Franquicia) en el que más jugó (si inactivo) o equipo en el que juega actualmente.
- team_city: Ciudad del equipo en el que más jugó (si inactivo) o equipo en el que juega actualmente.
- playercode: Esta columna informa del nombre + si el jugador es histórico.
- from_year: Año de inicio en la NBA
- to_year: Año de fin en la NBA
- dleague_flag: Indicador de si el jugador ha jugado en la G-League
- nba_flag: Indicador de si el jugador ha jugado en la NBA
- games_played_flag: Indicador de si el jugador ha disputado partidos
- draft_year: Año en que fue seleccionado en el draft
- draft_round: Ronda del draft en que fue seleccionado
- draft_number: Número de selección en el draft
- greatest_75_flag: Indicador de si el jugador forma parte de los 75 mejores de la historia de la NBA.


Esta tabla nos muestra información general de los jugadores. En este caso vamos a cambiar varios nombres de columnas de esta tabla y además vamos a eliminar las siguientes columnas:
- display_first_last: Ya que no aporta información relevante
- display_last_comma_first: Ya que no aporta información relevante
- display_fi_last: Ya que no aporta información relevante
- player_slug: Ya que no aporta información relevante
- playercode: Ya que no aporta información relevante
- last_affiliation: Ya que no aporta información relevante a nuestro estudio actualmente.

In [10]:
df_work = df_work.drop(columns= ['display_first_last', 'display_last_comma_first', 'display_fi_last', 'player_slug', 'playercode', 
                                 'last_affiliation'])
print ("Columnas eliminadas")

Columnas eliminadas


In [11]:
df_work = renombrar_columnas(df_work, 'school', 'organization_2')
df_work = renombrar_columnas(df_work, 'season_exp', 'seniority')
df_work = renombrar_columnas(df_work, 'rosterstatus', 'is_active_2')
df_work = renombrar_columnas(df_work, 'team_id', 'historic_team')
df_work = renombrar_columnas(df_work, 'team_city', 'historic_city')
df_work = renombrar_columnas(df_work, 'team_name', 'historic_team_name')
df_work = renombrar_columnas(df_work, 'team_code', 'historic_code')
df_work = renombrar_columnas(df_work, 'team_abbreviation', 'historic_team_abbrv')
df_work = renombrar_columnas(df_work, 'dleague_flag', 'g-league')
df_work = renombrar_columnas(df_work, 'nba_flag', 'nba')
df_work.columns

Nombre de la columna cambiado de school a organization_2
Nombre de la columna cambiado de season_exp a seniority
Nombre de la columna cambiado de rosterstatus a is_active_2
Nombre de la columna cambiado de team_id a historic_team
Nombre de la columna cambiado de team_city a historic_city
Nombre de la columna cambiado de team_name a historic_team_name
Nombre de la columna cambiado de team_code a historic_code
Nombre de la columna cambiado de team_abbreviation a historic_team_abbrv
Nombre de la columna cambiado de dleague_flag a g-league
Nombre de la columna cambiado de nba_flag a nba


Index(['id', 'jersey_num', 'general_team', 'general_city', 'general_team_name',
       'general_team_abbv', 'full_name', 'is_active', 'player_name', 'season',
       'round_number', 'round_pick', 'overall_pick', 'draft_type',
       'draft_team', 'draft_city', 'draft_team_name', 'draft_team_abbrv',
       'organization', 'organization_type', 'first_name', 'last_name',
       'birthdate', 'organization_2', 'country', 'height', 'weight',
       'seniority', 'jersey', 'position', 'is_active_2',
       'games_played_current_season_flag', 'historic_team',
       'historic_team_name', 'historic_team_abbrv', 'historic_code',
       'historic_city', 'from_year', 'to_year', 'g-league', 'nba',
       'games_played_flag', 'draft_year', 'draft_round', 'draft_number',
       'greatest_75_flag'],
      dtype='str')

***Paso 2 --> Gestión de nulos***

Tras esta primera limpieza, vamos a revisar la cantidad de nulos que hay por columna en busca de columnas con una gran cantidad de nulos que puedan afectar al resultado final del estudio: 

In [12]:
nulos = nulos_tabla(df_work)

Como no podemos visualizarlo todo vamos a exportarlo para visualizarlo en excel:

In [13]:
nulos.to_csv("nulos.csv", index= False)

Una vez exportado, vamos a comenzar a tratar las columnas: 

- Columnas de general y historic teams

Estas columnas parece que informan de lo mismo por lo que vamos a eliminar las columnas de historic ya que contienen nulos mientras que las de general no.

In [14]:
df_work.columns

Index(['id', 'jersey_num', 'general_team', 'general_city', 'general_team_name',
       'general_team_abbv', 'full_name', 'is_active', 'player_name', 'season',
       'round_number', 'round_pick', 'overall_pick', 'draft_type',
       'draft_team', 'draft_city', 'draft_team_name', 'draft_team_abbrv',
       'organization', 'organization_type', 'first_name', 'last_name',
       'birthdate', 'organization_2', 'country', 'height', 'weight',
       'seniority', 'jersey', 'position', 'is_active_2',
       'games_played_current_season_flag', 'historic_team',
       'historic_team_name', 'historic_team_abbrv', 'historic_code',
       'historic_city', 'from_year', 'to_year', 'g-league', 'nba',
       'games_played_flag', 'draft_year', 'draft_round', 'draft_number',
       'greatest_75_flag'],
      dtype='str')

In [15]:
df_work = df_work.drop(columns = ['historic_team', 'historic_team_name', 'historic_team_abbrv', 'historic_code', 'historic_city'])
print ("Columnas eliminadas")

Columnas eliminadas


- Columnas de jersey

*1. Combinación de columnas para reducir nulos*

Una de las columnas presenta un 31.55% de nulos mientras que la otra un 0.04% antes de eliminar la primera vamos a ver si podemos reutilizar
los valores de dicha columna para cubrir ese 0.04% y bajar el total de nulos para ello nos apoyaremos en una función predefinida.

In [16]:
precombine_registers(df_work, 'jersey', 'jersey_num')

jersey tiene 75430 registros
jersey_num tiene 110148 registros
Hay 75399 columnas con registros en ambas columnas


'12 carecen de registros'

Una vez identificadas podemos combinar los registros de las dos para así reducir el total de nulos de ambas. Para hacerlo vamos a utilizar la función combine.

In [17]:
df_work['jersey_def'] = df_work['jersey_num'].combine_first(df_work['jersey'])

A continuación obtenemos el total de nulos de la nueva columna para comprobar que efectivamente se han combinado correctamente:

In [18]:
df_work['jersey_def'].isna().sum()

np.int64(12)

Como podemos observar la combinación de ambas columnas ha sido exitosa por lo que eliminamos las columnas previas y nos quedamos con esta última que acabamos de crear.

In [19]:
df_work = df_work.drop(columns = ['jersey_num', 'jersey'])
print ("Columnas eliminadas")

Columnas eliminadas


*2. Tratamiento de nulos*

En primer lugar vamos a ver la cantidad de nulos que quedan: 

In [20]:
df_work['jersey_def'].isnull().sum()

np.int64(12)

Las 12 columnas restante las vamos a eliminar ya que no impactará en el resultado final eliminar tan pocas columnas.

In [21]:
#Eliminamos las filas de jugadores sin dorsal
df_work = eliminar_nulos(df_work, 'jersey_def')

El dataframe tiene la siguiente forma (110191, 40)
Se han eliminado los registros
El dataframe tiene la siguiente forma tras el borrado (110179, 40)


- Columnas organization

En este caso, vamos a proceder como hemos hecho en el caso anterior ya que tenemos una columna organization con un 26.06% de nulos y otra con un 14.96%. Por lo que realizaremos el mismo trabajo:

*1. Combinación de columnas*

In [22]:
precombine_registers(df_work, 'organization', 'organization_2')

organization tiene 81462 registros
organization_2 tiene 93700 registros
Hay 69683 columnas con registros en ambas columnas


'4700 carecen de registros'

Como en el caso anterior, podemos combinar columnas para reducir el numero de nulos totales en ambas columnas por lo que procedemos a realizar el mismo trabajo creando una nueva columna.

In [23]:
df_work['organization_def'] = df_work['organization_2'].combine_first(df_work['organization'])

Ahora comprobamos el numero de nulos en la nueva columna para ver que se han combinado correctamente:

In [24]:
df_work['organization_def'].isna().sum()

np.int64(4700)

Como podemos observar la combinación de ambas columnas ha sido exitosa por lo que eliminamos las columnas previas y nos quedamos con esta última que acabamos de crear.

In [25]:
df_work = df_work.drop(columns = ['organization_2', 'organization'])
print ("Columnas eliminadas")

Columnas eliminadas


*2. Limpieza de nulos*

En primer lugar, vamos a mirar la cantidad de nulos de nuevo y el porcentaje que representan:

In [26]:
nulos_tabla(df_work[['organization_def']])

,columna,total_nulos,porcentaje
0,organization_def,4700,4.27


Como podemos aún tenemos un 4.27% de nulos en la columna. Al tratarse de una columna categórica, vamos a sustituir los valores por 'Unknown'.

In [27]:
df_work['organization_def'] = df_work['organization_def'].fillna('Unknown')

Comprobamos que el cambio se ha realizado correctamente volviendo a obtener los nulos para la columna:

In [28]:
nulos_tabla(df_work[['organization_def']])

,columna,total_nulos,porcentaje
0,organization_def,0,0.0


- Columna organization type

Esta es una columna categórica que contine un 26.06% de nulos y no se puede extrapolar a ninguna otra para reducir el número de nulos por ello, esta columna requiere de otro tipo de tratamiento.
En este caso, al haber tal número de nulos y ser una columna categórica lo que vamos a realizar es crear una nueva categoría en dicha columna que se llama "Unknown" y que sustituirá a los valores nulos.

In [29]:
#Sustituimos los valores nulos por Unknown
df_work['organization_type'] = df_work['organization_type'].fillna('Unknown')
print("Sustitución realizada")

Sustitución realizada


Obtenemos de nuevo los nulos para ver que no quedan valores nulos:

In [30]:
#Comprobamos que no quedan nulos en la columna
nulos_tabla(df_work[['organization_type']])

,columna,total_nulos,porcentaje
0,organization_type,0,0.0


- Columnas de nombres

Como podemos ver, actualmente hay varias columnas que informan de nombres de jugadores. Las columnas son las siguientes: 

    1. player _name 25.98%
    2. first_name 14.45%
    3. last_name 14.45%
    4. full_name 0.58%

Una destaca por tener muchos menos nulos que el resto pero como en el caso anterior vamos a ver si podemos reducir el numero de nulos de esta última columna con los registros de las otras.

*1. Combinación de columnas*

In [31]:
#Lo primero que vamos a realizar es ver los registros en cada columna
df_work[['player_name', 'first_name', 'last_name', 'full_name']]

,player_name,first_name,last_name,full_name
0,Bobby Phills,NaN,NaN,Bobby Phills
1,Will Perdue,NaN,NaN,Will Perdue
2,Steven Smith,Steven,Smith,Steven Smith
3,Eddie Johnson,Eddie,Johnson,Eddie Johnson
4,Clifford Rozier,Clifford,Rozier,Clifford Rozier
...,...,...,...,...
110186,Zion Williamson,Zion,Williamson,Zion Williamson
110187,Stephen Curry,Stephen,Curry,Stephen Curry
110188,Kevin Durant,Kevin,Durant,Kevin Durant
110189,Zion Williamson,Zion,Williamson,Zion Williamson


Vamos a combinar los registros de la columnas first_name y last_name para crear una columna nueva de nombres completos y después poder extrapolar desde ella.

In [32]:
df_work['player_name_2'] = df_work['first_name']+ ' ' + df_work['last_name']
print ("Columnas combinadas")

Columnas combinadas


Volvemos a revisar las columnas: 

In [33]:
df_work[['player_name', 'player_name_2', 'full_name']]

,player_name,player_name_2,full_name
0,Bobby Phills,NaN,Bobby Phills
1,Will Perdue,NaN,Will Perdue
2,Steven Smith,Steven Smith,Steven Smith
3,Eddie Johnson,Eddie Johnson,Eddie Johnson
4,Clifford Rozier,Clifford Rozier,Clifford Rozier
...,...,...,...
110186,Zion Williamson,Zion Williamson,Zion Williamson
110187,Stephen Curry,Stephen Curry,Stephen Curry
110188,Kevin Durant,Kevin Durant,Kevin Durant
110189,Zion Williamson,Zion Williamson,Zion Williamson


Ahora procedemos a combinar los registros

**Full_name + player_name**

In [34]:
#Primero revisamos los nulos, registros en ambas y columnas sin registros.
precombine_registers(df_work, 'player_name', 'full_name')

player_name tiene 81558 registros
full_name tiene 109537 registros
Hay 81415 columnas con registros en ambas columnas


'499 carecen de registros'

In [35]:
#Combinamos ambas columnas 
df_work['full_name'] = df_work['full_name'].combine_first(df_work['player_name'])

In [36]:
#Comprobamos el número de nulos 
df_work['full_name'].isnull().sum()

np.int64(499)

**Full_name + player_name 2**

In [37]:
#Primero revisamos los nulos, registros en ambas y columnas sin registros.
precombine_registers(df_work, 'player_name_2', 'full_name')

player_name_2 tiene 94268 registros
full_name tiene 109680 registros
Hay 94268 columnas con registros en ambas columnas


'499 carecen de registros'

Como podemos ver en este caso las filas sin registros ya son las mismas que teniamos tras juntar las dos previas columnas por lo que no tiene sentido que juntemos estas columnas. Por lo que ahora nos deshazemos de las columnas que ya no nos sirven:

In [38]:
df_work = df_work.drop(columns = ['player_name', 'first_name', 'last_name', 'player_name_2'])
print("Columnas eliminadas")

Columnas eliminadas


*2. Tratamiento de nulos*

Al no tener el nombre del jugador esos registros no nos sirven por lo que vamos a eliminarlos del dataset ya que ya sabemos que no nos sirven.

In [39]:
df_work = eliminar_nulos(df_work,'full_name')

El dataframe tiene la siguiente forma (110179, 36)
Se han eliminado los registros
El dataframe tiene la siguiente forma tras el borrado (109680, 36)


- Columna season

*1. Combinamos columnas*

La columna season informa de la temporada en la que se celebró el draft en el que el jugador fue drafteado. Esta columna presenta un 25.98% de nulos mientras que la columna draft_year hay un 14.45% por lo que vamos a combinar ambas columnas para ver si podemos reducir el numero de nulos.

In [40]:
#En primer lugar, realizamos el estudio de nulos por columnas antes de combinarlas
precombine_registers(df_work, 'draft_year', 'season')

draft_year tiene 94268 registros
season tiene 81558 registros
Hay 70133 columnas con registros en ambas columnas


'3987 carecen de registros'

In [41]:
#Combinamos ambas columnas 
df_work['draft_year'] = df_work['draft_year'].combine_first(df_work['season'])

In [42]:
#Revisamos el número de nulos para comprobar que se han fusionado correctamente
df_work['draft_year'].isna().sum()

np.int64(3987)

In [43]:
#Eliminamos la columna season 
df_work = df_work.drop(columns=['season'])
print("Columna eliminada")

Columna eliminada


*2. Tratamiento de nulos*

En este caso, para tratar los nulos, vamos a revisar en primer lugar los valores informados en la columna draft_year para ver si ya tiene en cuenta a los jugadores no drafteados.

In [44]:
df_work['draft_year'].unique()

array([1991.0, 1988.0, '1991', '1981', '1994', '1985', '1990', '1993',
       '1992', 1992.0, 'Undrafted', 1983.0, '1987', 1995.0, 1985.0,
       '1996', '1978', 1990.0, '1986', '1988', '1997', '1995', 1993.0,
       1997.0, '1989', '1983', '1998', 1980.0, 1999.0, 1998.0, '1999',
       '2000', 1994.0, '1984', 2000.0, 2001.0, '2001', 1987.0, nan,
       '2002', 2002.0, '2003', 2003.0, '2004', 1996.0, '2005', 2004.0,
       2005.0, '2006', 2006.0, 2007.0, '2007', '2008', 2008.0, 2009.0,
       '2009', '2010', 2010.0, '2011', 2011.0, '1982', '2012', '2013',
       2013.0, 2012.0, '2014', '2015', 2015.0, '2016', 2016.0, '2017',
       2017.0, '2018', 2018.0, '2019', 2019.0, '2020', 2020.0, 2021.0,
       '2021', 2022.0, '2022'], dtype=object)

Como podemos ver, ya tiene en cuenta a los jugadores no drafteados. Lo siguiente que vamos a ver es el porcentaje de nulos remanente en la columna:

In [45]:
nulos_tabla(df_work[['draft_year']])

,columna,total_nulos,porcentaje
0,draft_year,3987,3.64


Al ser tan pequeño el porcentaje remanente y visto que no tenemos información de si esos jugadores fueron drafteados, rellenamos los nulos con el valor Undrafted.

In [46]:
df_work['draft_year'] = df_work['draft_year'].fillna('Undrafted')
print("Sustitución realizada")

Sustitución realizada


Por último, revisamos de nuevo el número de nulos: 

In [47]:
nulos_tabla(df_work[['draft_year']])

,columna,total_nulos,porcentaje
0,draft_year,0,0.0


- Columnas round_number y round_pick

*1. Combinación de columnas*

En este caso ocurre lo mismo que en el anterior, la columna round_number presenta un 25.98% de nulos y su equivalente; la columna draft round presenta un 15.52% por lo que como antes las combinaremos para reducir el número de nulos. 

In [48]:
#En primer lugar, realizamos el estudio de nulos por columnas antes de combinarlas
precombine_registers(df_work, 'round_number', 'draft_round')

round_number tiene 81558 registros
draft_round tiene 93084 registros
Hay 70133 columnas con registros en ambas columnas


'5171 carecen de registros'

In [49]:
#Combinamos ambas columnas 
df_work['draft_round'] = df_work['draft_round'].combine_first(df_work['round_number'])

In [50]:
#Revisamos el número de nulos para comprobar que se han fusionado correctamente
df_work['draft_round'].isna().sum()

np.int64(5171)

In [51]:
#Eliminamos la columna round_number
df_work = df_work.drop(columns=['round_number'])
print("Columna eliminada")

Columna eliminada


A continuación, repetimos el mismo trabajo pero con las columnas round_pick con un 25.98% y la columna draft_number:

In [52]:
#En primer lugar, realizamos el estudio de nulos por columnas antes de combinarlas
precombine_registers(df_work, 'round_pick', 'draft_number')

round_pick tiene 81558 registros
draft_number tiene 93022 registros
Hay 70133 columnas con registros en ambas columnas


'5233 carecen de registros'

In [53]:
#Combinamos ambas columnas 
df_work['draft_number'] = df_work['draft_number'].combine_first(df_work['round_pick'])

In [54]:
#Revisamos el número de nulos para comprobar que se han fusionado correctamente
df_work['draft_number'].isna().sum()

np.int64(5233)

In [55]:
#Eliminamos la columna round_number
df_work = df_work.drop(columns=['round_pick'])
print("Columna eliminada")

Columna eliminada


*2. Tratamiento de nulos*

**Columna draft_number**

Revisamos la cantidad de nulos en la columna:

In [56]:
nulos_tabla(df_work[['draft_number']])

,columna,total_nulos,porcentaje
0,draft_number,5233,4.77


Tenemos un 4.77% de nulos en la columna draft_number. En este caso, vamos a ver primero cuantos de esos jugadores son Undrafted comparando con la columna draft_year.

In [57]:
No_drafteados = df_work[df_work['draft_number'].isna() & (df_work['draft_year'] == 'Undrafted')]
print(f'Registros : {len(No_drafteados)}')

Registros : 3987


Como podemos observar, hay registros que son jugadores no drafteados por lo que vamos a sustituir esos valores por el valor Undrafted.

In [58]:
#Para realizar la sustitución de dichos valores vamos a utilizar la función loc.
#esta función nos permite seleccionar y modificar filas y columnas por etiquetas.

df_work.loc[df_work['draft_year'] == 'Undrafted', 'draft_number'] = 'Undrafted'
print ("Valores sustituidos")

Valores sustituidos


Ahora volvemos a ver la cantidad de nulos en la columna para ver que se han modificado correctamente.

In [59]:
nulos_tabla(df_work[['draft_number']])

,columna,total_nulos,porcentaje
0,draft_number,1246,1.14


Como podemos ver la sustitución ha sido existosa ya que el remanente de nulos son la resta de los anteriores menos los valores cambiados. A continuación, vamos a eliminar el resto de nulos de la columna.

In [60]:
df_work = eliminar_nulos(df_work, 'draft_number')

El dataframe tiene la siguiente forma (109680, 33)
Se han eliminado los registros
El dataframe tiene la siguiente forma tras el borrado (108434, 33)


**Columna draft_round**

Revisamos la cantidad de nulos en la columna: 

In [61]:
nulos_tabla(df_work[['draft_round']])

,columna,total_nulos,porcentaje
0,draft_round,3987,3.68


Cruzamos la cantidad de nulos con los jugadores no drafteados:

In [62]:
No_drafteados_2 = df_work[df_work['draft_round'].isna() & (df_work['draft_year'] == 'Undrafted')]
print(f'Registros : {len(No_drafteados_2)}')

Registros : 3987


Hacemos ahora la misma sustitución con loc. y esto debería eliminar todos los nulos de la columna:

In [63]:
df_work.loc[df_work['draft_year'] == 'Undrafted', 'draft_round'] = 'Undrafted'
print ("Valores sustituidos")

Valores sustituidos


Comprobamos el resultado obteniendo de nuevo los nulos:

In [64]:
nulos_tabla(df_work[['draft_round']])

,columna,total_nulos,porcentaje
0,draft_round,0,0.0


- Columna overall_pick

La columna overall_pick nos muestra el pick general del draft que fue el jugador. Esta columna a parte de tener un 25.98% de nulos presenta datos que extrapolaremos de las columnas que hemos calculado anteriormente. 
Por esto, vamos a eliminar esta columna y vamos a generar una nueva con el mismo propósito durante la fase de tratamiento ya que las dos columnas mencionadas anteriormente requieren de cierto tratamiento.

In [65]:
#Eliminamos la columna
df_work = df_work.drop(columns=['overall_pick'])
print("Columna eliminada")

Columna eliminada


- Columna draft_type 

Esta columna representa el tipo de draft en el que el jugador fue seleccionado visto que requerimos de mayor entendimiento del funcionamiento de esta columna, vamos a ver los valores que alberga antes de tomar una decisión.

In [66]:
df_work['draft_type'].unique()

<StringArray>
['Draft', nan]
Length: 2, dtype: str

Visto que la columna solo tiene dos tipos de registro que no aportan valor y que presenta un 25.98% de nulos, vamos a eliminar la columna. 

In [67]:
df_work = df_work.drop(columns = ['draft_type'])
print ("Columna eliminada")

Columna eliminada


- Columna draft_team

Esta columna contiene el id del equipo que drafteó al jugador por lo que si el jugador no fue drafteado la columna estará en blanco. Vamos a revisar el total de nulos en la columna:

In [68]:
nulos_tabla(df_work[['draft_team']])

,columna,total_nulos,porcentaje
0,draft_team,26876,24.79


Vamos a ver de este total de nulos los que se deben a que el jugador no fue drafteado:

In [69]:
No_drafteados_3 = df_work[df_work['draft_team'].isna() & (df_work['draft_year'] == 'Undrafted')]
print(f'Registros : {len(No_drafteados_3)}')

Registros : 22284


Ahora vamos a sustituir esos valores por 0 al tratarse esta columna una columna de tipo float. Durante la fase de transformación, gestionaremos estos valores de 0 si fuera necesario:

In [70]:
df_work.loc[df_work['draft_year'] == 'Undrafted', 'draft_team'] = 0
print ("Valores sustituidos")

Valores sustituidos


Volvemos a revisar ahora el total de nulos en la columna y el porcentaje que representan:

In [71]:
nulos_tabla(df_work[['draft_team']])

,columna,total_nulos,porcentaje
0,draft_team,4592,4.23


Estos nulos cuyo origen es desconocido vamos a eliminarlos del dataset ya que pueden causar alucinaciones en nuestro estudio.

In [72]:
df_work = eliminar_nulos(df_work, 'draft_team')

El dataframe tiene la siguiente forma (108434, 31)
Se han eliminado los registros
El dataframe tiene la siguiente forma tras el borrado (103842, 31)


- Columna draft_city

Esta columna dice la ciudad del equipo en el que el jugador fue drafteado. Lo primero que vamos a hacer como en el caso anterior es revisar el total de nulos en la columna:

In [73]:
nulos_tabla(df_work[['draft_city']])

,columna,total_nulos,porcentaje
0,draft_city,22284,21.46


Ahora vamos a revisar de estos cuantos se deben a que el jugador es no drafteado que intuimos que será un 100%.

In [74]:
No_drafteados_4 = df_work[df_work['draft_city'].isna() & (df_work['draft_year'] == 'Undrafted')]
print(f'Registros : {len(No_drafteados_4)}')

Registros : 22284


Dado que el total de nulos se debe a que los jugadores no fueron drafteados los sustituimos por el valor undrafted:

In [75]:
df_work.loc[df_work['draft_year'] == 'Undrafted', 'draft_city'] = 'Undrafted'
print ("Valores sustituidos")

Valores sustituidos


Y volvemos a revisar los nulos:

In [76]:
nulos_tabla(df_work[['draft_city']])

,columna,total_nulos,porcentaje
0,draft_city,0,0.0


- Columna Draft_team_name

Esta columna es como el caso anterior por lo que procedemos a realizar el mismo trabajo para eliminar los nulos. En primer lugar, obtenemos el total de registros nulos:

In [77]:
nulos_tabla(df_work[['draft_team_name']])

,columna,total_nulos,porcentaje
0,draft_team_name,22284,21.46


Comprobamos que provengan de jugadores no drafteados:

In [78]:
No_drafteados_5 = df_work[df_work['draft_team_name'].isna() & (df_work['draft_year'] == 'Undrafted')]
print(f'Registros : {len(No_drafteados_5)}')

Registros : 22284


Sustituimos:

In [79]:
df_work.loc[df_work['draft_year'] == 'Undrafted', 'draft_team_name'] = 'Undrafted'
print ("Valores sustituidos")

Valores sustituidos


Comprobamos:

In [80]:
nulos_tabla(df_work[['draft_team_name']])

,columna,total_nulos,porcentaje
0,draft_team_name,0,0.0


- Columna draft_team_abbrv

Esta es otra columna como el caso anterior por lo que realizamos de nuevo el mismo trabajo. Comprobamos los nulos:

In [81]:
nulos_tabla(df_work[['draft_team_abbrv']])

,columna,total_nulos,porcentaje
0,draft_team_abbrv,22284,21.46


Comprobamos que provienen de jugadores no drafteados:

In [82]:
No_drafteados_6 = df_work[df_work['draft_team_abbrv'].isna() & (df_work['draft_year'] == 'Undrafted')]
print(f'Registros : {len(No_drafteados_6)}')

Registros : 22284


Sustituimos:

In [83]:
df_work.loc[df_work['draft_year'] == 'Undrafted', 'draft_team_abbrv'] = 'Undrafted'
print ("Valores sustituidos")

Valores sustituidos


Comprobamos los nulos de nuevo:

In [84]:
nulos_tabla(df_work[['draft_team_abbrv']])

,columna,total_nulos,porcentaje
0,draft_team_abbrv,0,0.0


- Columna height

La columna height informa de la altura de los jugadores por lo que es una columna que nos hará falta para este estudio. En primer lugar antes de tratarla, vamos a ver la cantidad de nulos que presenta:

In [85]:
nulos_tabla(df_work[['height']])

,columna,total_nulos,porcentaje
0,height,15961,15.37


*1. Transformación de valores*

Lo primero que vamos a revisar es como está informada esta columna ya que el tipo de columna que presenta es de string.

In [86]:
df_work['height'].unique()

<StringArray>
[     nan, '06-Aug', '06-Nov',    '7-0', '06-Sep', '07-Jun', '06-May',
 '06-Jul', '06-Oct', '06-Jun', '05-Nov', '06-Mar', '06-Apr', '06-Feb',
 '06-Jan',    '6-0', '07-Feb', '07-Jul', '05-Oct', '07-Jan', '07-Mar',
 '07-May', '05-May', '05-Sep', '07-Apr']
Length: 25, dtype: str

Como podemos observar, pandas ha registrado la columna que venía en pies y pulgadas como fecha por lo que antes de tratar los nulos, vamos a tener que transformarla en centímetros para que sea más entendible.
Lo primero que vamos a hacer es convertirla a pies y pulgadas:

In [87]:
#Para convertirlo vamos a generar un diccionario en primer lugar con las conversiones
altura_map = {
    # 5 pies
    '5-0': "5' 0' '",
    '05-Jan': "5' 1''", '05-Feb': "5' 2''", '05-Mar': "5' 3''",
    '05-Apr': "5' 4''", '05-May': "5' 5''", '05-Jun': "5' 6''",
    '05-Jul': "5' 7''", '05-Aug': "5' 8''", '05-Sep': "5' 9''",
    '05-Oct': "5' 10''", '05-Nov': "5' 11''",

    # 6 pies
    '6-0': "6' 0''",
    '06-Jan': "6' 1''", '06-Feb': "6' 2''", '06-Mar': "6' 3''",
    '06-Apr': "6' 4''", '06-May': "6' 5''", '06-Jun': "6' 6''",
    '06-Jul': "6' 7''", '06-Aug': "6' 8''", '06-Sep': "6' 9''",
    '06-Oct': "6' 10''", '06-Nov': "6' 11''",

    # 7 pies
    '7-0': "7' 0''",
    '07-Jan': "7' 1''", '07-Feb': "7' 2''", '07-Mar': "7' 3''",
    '07-Apr': "7' 4''", '07-May': "7' 5''", '07-Jun': "7' 6''",
    '07-Jul': "7' 7''", '07-Aug': "7' 8''", '07-Sep': "7' 9''",
    '07-Oct': "7' 10''", '07-Nov': "7' 11''",
}

#A continuación con la función map corregimos los valores de la columna height
df_work['height_def'] = df_work['height'].map(altura_map)
print ("Columna corregida")

Columna corregida


In [88]:
#Revisamos como han quedado los valores de la columna
df_work['height_def'].unique()

<StringArray>
[      nan,  '6' 8''', '6' 11''',  '7' 0''',  '6' 9''',  '7' 6''',  '6' 5''',
  '6' 7''', '6' 10''',  '6' 6''', '5' 11''',  '6' 3''',  '6' 4''',  '6' 2''',
  '6' 1''',  '6' 0''',  '7' 2''',  '7' 7''', '5' 10''',  '7' 1''',  '7' 3''',
  '7' 5''',  '5' 5''',  '5' 9''',  '7' 4''']
Length: 25, dtype: str

Para pasar de inches a centimetros vamos a utilizar una función:

In [89]:
df_work['height_definitivo'] = df_work['height_def'].apply(convertir_altura)

Revisamos la conversión de valores:

In [90]:
df_work['height_definitivo'].unique()

array([   nan, 203.2 , 210.82, 213.36, 205.74, 228.6 , 195.58, 200.66,
       208.28, 198.12, 180.34, 190.5 , 193.04, 187.96, 185.42, 182.88,
       218.44, 231.14, 177.8 , 215.9 , 220.98, 226.06, 165.1 , 175.26,
       223.52])

*2. Tratamiento de nulos*

Una vez realizada la transformación ahora vamos a gestionar los nulos. Esta columna va a tener un alto impacto en nuestro estudio ya que una de las cosas que queremos estudiar es el impacto que tiene la altura en la posición de un jugador. 
Por ello y para evitar sesgos en el estudio, vamos a deshacernos de las filas a las que les falte la altura ya que la muestra seguirá siendo suficientemente grande.

In [91]:
df_work = eliminar_nulos(df_work, 'height_definitivo')

El dataframe tiene la siguiente forma (103842, 33)
Se han eliminado los registros
El dataframe tiene la siguiente forma tras el borrado (87881, 33)


Por último, borramos las columnas que ya no nos sirven y renombramos esta última.

In [92]:
df_work = df_work.drop(columns = ['height', 'height_def'])
print ("Columnas eliminadas")

Columnas eliminadas


- Columna weight

Esta columna informa del peso del jugador, va a tener un alto impacto en el estudio por lo que comentabamos anteriormente. Por lo que requiere de alta precisión. Para empezar a tratar la columna vamos a ver la cantidad de nulos en la columna:

In [93]:
nulos_tabla(df_work[['weight']])

,columna,total_nulos,porcentaje
0,weight,0,0.0


Como podemos observar, la columna ya no presenta nulos por lo que proseguimos con la limpieza.

- Columna position

Como en el caso anterior, esta columna que informa de la posición del jugador tiene un alto impacto en el estudio. Es por ello por lo que requiere de mucha precisión en el estudio, para ello vamos a revisar los nulos en la columna.

In [94]:
nulos_tabla(df_work[['position']])

,columna,total_nulos,porcentaje
0,position,0,0.0


Como podemos observar, la columna ya no presenta nulos por lo que proseguimos con la limpieza.

Según parece, el dataframe está ya limpio de nulos pero vamos a confirmarlo:

In [95]:
#Extraemos los valores nulos de todo el datafram
nulos_2 = nulos_tabla(df_work)
#Lo exportamos para visualizarlo
nulos_2.to_csv("nulos2.csv", index=False)

En efecto parece que el dataframe está ya limpio de nulos por lo que vamos a exportarlo y vamos a hacer en un segundo script de limpieza el tratamiento de duplicados.
Antes de eliminarlo, vamos a revisar las columnas existentes en nuestro dataframe y eliminar las columnas duplicadas.

In [96]:
df_work.columns

Index(['id', 'general_team', 'general_city', 'general_team_name',
       'general_team_abbv', 'full_name', 'is_active', 'draft_team',
       'draft_city', 'draft_team_name', 'draft_team_abbrv',
       'organization_type', 'birthdate', 'country', 'weight', 'seniority',
       'position', 'is_active_2', 'games_played_current_season_flag',
       'from_year', 'to_year', 'g-league', 'nba', 'games_played_flag',
       'draft_year', 'draft_round', 'draft_number', 'greatest_75_flag',
       'jersey_def', 'organization_def', 'height_definitivo'],
      dtype='str')

In [97]:
#Vamos a deshacernos de las columnas duplicadas e inservibles a nuestro estudio.
df_work = df_work.drop(columns = ['is_active_2', 'games_played_current_season_flag','games_played_flag'])
print ("Columnas eliminadas")

Columnas eliminadas


In [ ]:
df_work.to_csv("Dataframe_sin_nulos.csv", index=False)